src/lib/libutil/check_expire.3
.\"	$OpenBSD: check_expire.3,v 1.9 2013/08/14 08:39:26 jmc Exp $
.\"
.\" Copyright (c) 2000 Todd C. Miller <Todd.Miller@courtesan.com>
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd $Mdocdate: August 14 2013 $
.Dt LOGIN_CHECK_EXPIRE 3
.Os
.Sh NAME
.Nm login_check_expire
.Nd check for password expiration
.Sh SYNOPSIS
.In stdio.h
.In util.h
.Ft int
.Fn login_check_expire "FILE *back" "struct passwd *pwd" "char *class" "int lastchance"
.Sh DESCRIPTION
The
.Fn login_check_expire
function is called by a
.Bx
Authentication login script to
check whether the user's password entry, as described by
.Fa pwd ,
has expired.
.Pp
If a
.Fa class
is specified, it is used instead of the class specified in the user's
password database entry.
If the
.Fa lastchance
argument is non-zero, the user's password has expired, and it has not been
expired longer than
.Dq password-dead
seconds (see
.Xr login.conf 5 ) ,
the user will be able to log in one last time to change the password.
.Sh RETURN VALUES
The
.Fn login_check_expire
function returns 0 if the user's password has not expired, and 1 if it has
expired or if an error occurred.
.br
Status and error messages are passed
back to the login script caller via the back channel,
.Fa back .
.Sh SEE ALSO
.Xr auth_subr 3 ,
.Xr authenticate 3 ,
.Xr login.conf 5


src/lib/libutil/check_expire.c
/*	$OpenBSD: check_expire.c,v 1.9 2013/04/29 00:19:19 okan Exp $	*/

/*
 * Copyright (c) 1997 Berkeley Software Design, Inc. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by Berkeley Software Design,
 *	Inc.
 * 4. The name of Berkeley Software Design, Inc.  may not be used to endorse
 *    or promote products derived from this software without specific prior
 *    written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY BERKELEY SOFTWARE DESIGN, INC. ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL BERKELEY SOFTWARE DESIGN, INC. BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 *	BSDI $From: check_expire.c,v 2.1 1997/08/08 18:38:25 prb Exp $
 */

#include <sys/types.h>

#include <errno.h>
#include <fcntl.h>
#include <pwd.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <syslog.h>
#include <time.h>
#include <tzfile.h>
#include <login_cap.h>
#include <bsd_auth.h>

#include "util.h"

static char *pwd_update(const struct passwd *, const struct passwd *);

int
login_check_expire(FILE *back, struct passwd *pwd, char *class, int lastchance)
{
	auth_session_t *as;
	login_cap_t *lc;
	quad_t dead, expire, warn;
	char *p;

	if ((as = auth_open()) == NULL) {
		syslog(LOG_ERR, "failed to create auth session");
		return (1);
	}
	if (auth_setpwd(as, pwd) < 0) {
		syslog(LOG_ERR, "failed to set pwd entry in auth session");
		return (1);
	}

	expire = auth_check_change(as);
	auth_close(as);

	if (expire != 0) {
		fprintf(back, BI_VALUE " expire %qd\n", expire);

		if (class == NULL)
			class = pwd->pw_class;

		if ((lc = login_getclass(class)) == NULL) {
			dead = 0;
			warn = 0;
		} else {
			dead = login_getcaptime(lc, "password-dead", 0, 0);
			warn = login_getcaptime(lc, "password-warn",
			    2 * DAYSPERWEEK * SECSPERDAY,
			    2 * DAYSPERWEEK * SECSPERDAY);
			if (dead < 0) {
				syslog(LOG_ERR, "class %s password-dead is %qd",
					lc->lc_class, dead);
				dead = 0;
			}
			if (warn < 0) {
				syslog(LOG_ERR, "class %s password-warn is %qd",
					lc->lc_class, warn);
				warn = 0;
			}
		}
		login_close(lc);

		/*
		 * If their password is dead (expired longer than
		 * password-dead) then just reject them.  If it is
		 * expired but not dead yet, reject them with a
		 * PWEXPIRED so login knows they can still sort of
		 * get in.
		 */
		if (expire < -dead) {
			syslog(LOG_WARNING, "%s: dead password", pwd->pw_name);
			fprintf(back, BI_VALUE
			    " errormsg Your password has expired\n");
			fprintf(back, BI_REJECT "\n");
			return (1);
		}
		if (expire < 0) {
			if (lastchance) {
				struct passwd *npwd;

				endpwent();

				/*
				 * Only let them play this game once.
				 * Set their password change time to 1.
				 * This will most certainly cause any
				 * expired password to be dead, as well.
				 */
				npwd = pw_dup(pwd);
				npwd->pw_change = 1;
				p = pwd_update(npwd, pwd);
				explicit_bzero(npwd->pw_passwd,
				    strlen(npwd->pw_passwd));
				free(npwd);
				if (p != NULL) {
					fprintf(back, BI_VALUE " errormsg %s",
					    auth_mkvalue(p));
					fprintf(back, BI_REJECT "\n");
					return (1);
				}
			}
			syslog(LOG_WARNING, "%s: expired password", pwd->pw_name);
			fprintf(back, BI_VALUE
			    " errormsg Your password has expired\n");
			fprintf(back, BI_PWEXPIRED "\n");
			return (1);
		}

		/*
		 * If their password is not expired but is about to expire
		 * then warn them.
		 */
		if (expire <= warn) {
			fprintf(back, BI_VALUE
			    " warnmsg Your password expires on %s\n",
			    ctime(&pwd->pw_change));
		}
	}
	return (0);
}

static char *
pwd_update(const struct passwd *pwd, const struct passwd *opwd)
{
	int tfd, pfd;

	pw_init();
	tfd = pw_lock(0);
	if (tfd < 0) {
		if (errno == EEXIST)
			return("the passwd file is busy.");
		else
			return("can't open passwd temp file");
	}

	pfd = open(_PATH_MASTERPASSWD, O_RDONLY|O_CLOEXEC, 0);
	if (pfd < 0) {
		pw_abort();
		return(strerror(errno));
	}

	pw_copy(pfd, tfd, pwd, opwd);
	if (pw_mkdb(pwd->pw_name, 0) < 0) {
		pw_abort();
		return("unable to update password database");
	}

	return(NULL);
}
